File: MiniDOMWebCtrls/DMD/CRM/CRMUtils.cs
using System;
using DMD;

namespace minidom.Forms
{
    public partial class Utils
    {
        public sealed class CCRMUtilsClas
        {
            public string CreateElencoUffici(Anagrafica.CUfficio selValue)
            {
                CCollection<Anagrafica.CUfficio> uffici;
                var writer = new System.Text.StringBuilder();
                if (Anagrafica.Uffici.Module.UserCanDoAction("list"))
                {
                    uffici = Anagrafica.Uffici.GetPuntiOperativi();
                }
                else
                {
                    uffici = Sistema.Users.CurrentUser.Uffici;
                }

                foreach (Anagrafica.CUfficio item in uffici)
                {
                    writer.Append("<option value=\"");
                    writer.Append(Databases.GetID(item));
                    writer.Append("\" ");
                    if (Databases.GetID(selValue) == Databases.GetID(item))
                        writer.Append("selected");
                    writer.Append(">");
                    writer.Append(DMD.WebUtils.HtmlEncode(item.Nome));
                    writer.Append("</option>");
                }

                return writer.ToString();
            }

            public string CreateElencoOperatori(Sistema.CUser selValue, bool onlyValid = true)
            {
                Sistema.CModule module;
                bool canList, canListOwn;
                int itemID;
                int value;
                Sistema.CGroup grp;
                var ret = new System.Text.StringBuilder();
                bool t;
                module = Finanziaria.Pratiche.Module;
                canList = module.UserCanDoAction("list");
                canListOwn = module.UserCanDoAction("list_own");
                value = Databases.GetID(selValue);
                t = false;
                grp = Sistema.Groups.GetItemByName("CRM");
                if (canList)
                {
                    // Otteniamo il gruppo Finanziaria
                    if (grp is object)
                    {
                        foreach (Sistema.CUser user in grp.Members)
                        {
                            if (onlyValid == false || user.UserStato == Sistema.UserStatus.USER_ENABLED)
                            {
                                itemID = Databases.GetID(user);
                                t = t | value == itemID;
                                ret = ret.Append("<option value=\"");
                                ret.Append(itemID);
                                ret.Append("\" ");
                                if (value == itemID)
                                    ret.Append("selected");
                                ret.Append(">");
                                ret.Append(DMD.WebUtils.HtmlEncode(user.Nominativo));
                                ret.Append("</option>");
                            }
                        }
                    }
                }
                else if (grp is object)
                {
                    foreach (Sistema.CUser user in grp.Members)
                    {
                        if (Sistema.Users.CurrentUser.Uffici.SameOffice(user))
                        {
                            if (onlyValid == false || user.UserStato == Sistema.UserStatus.USER_ENABLED)
                            {
                                itemID = Databases.GetID(user);
                                t = t | value == itemID;
                                ret.Append("<option value=\"");
                                ret.Append(itemID);
                                if (value == itemID)
                                    ret.Append("selected");
                                ret.Append(">");
                                ret.Append(DMD.WebUtils.HtmlEncode(user.Nominativo));
                                ret.Append("</option>");
                            }
                        }
                    }
                }

                if (selValue is object & t == false)
                {
                    itemID = Databases.GetID(selValue);
                    if (value == itemID)
                    {
                        ret.Append("<option value=\"");
                        ret.Append(itemID);
                        ret.Append("\" selected>");
                        ret.Append(selValue.Nominativo);
                        ret.Append("</option>");
                    }
                    else
                    {
                        ret.Append("<option value=\"");
                        ret.Append(value);
                        ret.Append("\">INVALID (");
                        ret.Append(value);
                        ret.Append(")</option>");
                    }
                }

                return ret.ToString();
            }

            public string CreateElencoCategorie(string selValue)
            {
                var items = new[] { "Urgente", "Importante", "Normale", "Poco importante" };
                return SystemUtils.CreateElenco(items, selValue);
            }

            public string GetIcon(Anagrafica.CPersona cliente)
            {
                string iconURL = "";
                if (cliente is object)
                    iconURL = cliente.IconURL;
                if (string.IsNullOrEmpty(iconURL))
                    return Sistema.ApplicationContext.BaseURL + "/minidom/widgets/images/default.GIF";
                return Sistema.ApplicationContext.BaseURL + "/minidom/widgets/websvc/getthumb.aspx?p=" + iconURL + "&w=32&h=32";
            }

            public string FormatOfferta(Finanziaria.COffertaCQS offerta)
            {
                if (offerta is null)
                    return "NULL";
                string testo = "";
                testo += "<table>";
                testo += "<tr>";
                testo += "<td>";
                var prodotto = offerta.Prodotto;
                if (prodotto is null)
                {
                    testo += "<b>?</b>";
                }
                else if (prodotto.IdTipoContratto == "C")
                {
                    testo += "<b>CQS&nbsp;</b>";
                }
                else
                {
                    testo += "<b>PD&nbsp;</b>";
                }

                testo += "<span class=\"blue\">" + Sistema.Formats.FormatValuta(offerta.Rata) + " €</span> x <span class=\"blue\">" + offerta.Durata + "</span> = " + Sistema.Formats.FormatValuta(offerta.MontanteLordo) + " €";
                testo += "</td>";
                testo += "</tr>";
                testo += "<tr>";
                testo += "<td>";
                testo += "<b>Netto:</b> <span class=\"blue\">" + Sistema.Formats.FormatValuta(offerta.NettoRicavo) + " €</span>, <b>TAN:</b> <span class=\"blue\">" + Sistema.Formats.FormatPercentage(offerta.TAN, 3) + " %</span> <b>TAEG:</b> <span class=\"blue\">" + Sistema.Formats.FormatPercentage(offerta.TAEG) + " %</span>";
                testo += "</td>";
                testo += "</tr>";
                testo += "</table>";
                return testo;
            }

            public string GetNominativoPersona(Anagrafica.CPersona persona)
            {
                string ret = "";
                if (persona is object)
                    ret = persona.Nominativo;
                if (string.IsNullOrEmpty(ret))
                {
                    if (persona is null)
                    {
                        ret = "Sconosciuto";
                    }
                    else
                    {
                        ret = DMD.RunTime.vbTypeName(persona) + "[" + Databases.GetID(persona) + "]";
                    }
                }

                return ret;
            }

            public string PreparaTestoNotifica(string titolo, Anagrafica.CPersona persona, Sistema.CUser operatore = null, DateTime? data = default)
            {
                string testo;
                testo = "";
                testo += "<table>";
                testo += "<tr>";
                testo += "<td><img src=\"" + GetIcon(persona) + "\" style=\"width:32px;height:32px;\" alt=\"img\" /></td>";
                testo += "<td>";
                testo += "<b style=\"color:red;font-size:12pt;\">" + titolo + "</b></br/>";
                testo += "<a style=\"color:blue;font-size:12pt;font-weight:bold;\" href=\"#\" onclick=\"SystemUtils.EditItem(Anagrafica.Persone.GetItemById(" + Databases.GetID(persona) + ")); return false;\">" + DMD.WebUtils.HtmlEncode(GetNominativoPersona(persona)) + "<a>";
                testo += "</td>";
                testo += "</tr>";
                testo += "</table>";
                if (data.HasValue)
                    testo += "<b class=\"font11pt\">Data:</b> <span class\"font11pt blue\">" + Sistema.Formats.FormatUserDateTime(DMD.DateUtils.Now()) + "</span>" + DMD.Strings.vbNewLine;
                if (operatore is object)
                    testo += "<b class=\"font11pt\">Operatore:</b> <span class\"blue font11pt\">" + DMD.WebUtils.HtmlEncode(operatore.Nominativo) + "</span>" + DMD.Strings.vbNewLine;
                return testo;
            }

            public string PreparaTestoNotificaURL(string titolo, Anagrafica.CPersona persona, Sistema.CUser operatore = null, DateTime? data = default)
            {
                string testo;
                string baseURL = Sistema.ApplicationContext.BaseURL;
                string baseName = "frm" + Sistema.ApplicationContext.IDAziendaPrincipale;
                testo = "";
                testo += "<table>";
                testo += "<tr>";
                testo += "<td><img src=\"" + GetIcon(persona) + "\" style=\"width:32px;height:32px;\" alt=\"img\" /></td>";
                testo += "<td>";
                testo += "<b style=\"color:red;font-size:12pt;\">" + titolo + "</b></br/>";
                testo += "<a style=\"color:blue;font-size:12pt;font-weight:bold;\" target=\"" + baseName + "\" href=\"" + baseURL + "modAnagrafica.aspx?_a=edit&id=" + Databases.GetID(persona) + "\">" + DMD.WebUtils.HtmlEncode(GetNominativoPersona(persona)) + "<a>";
                testo += "</td>";
                testo += "</tr>";
                testo += "</table>";
                if (data.HasValue)
                    testo += "<b class=\"font11pt\">Data:</b> <span class\"font11pt blue\">" + Sistema.Formats.FormatUserDateTime(DMD.DateUtils.Now()) + "</span>" + DMD.Strings.vbNewLine;
                if (operatore is object)
                    testo += "<b class=\"font11pt\">Operatore:</b> <span class\"blue font11pt\">" + DMD.WebUtils.HtmlEncode(operatore.Nominativo) + "</span>" + DMD.Strings.vbNewLine;
                return testo;
            }
        }

        private static CCRMUtilsClas m_CRMUtils = null;

        public static CCRMUtilsClas CRMUtils
        {
            get
            {
                if (m_CRMUtils is null)
                    m_CRMUtils = new CCRMUtilsClas();
                return m_CRMUtils;
            }
        }
    }
}

File: MiniDOMWebCtrls/DMD/CRM/CRMUtils.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.WebSite
Imports minidom.CustomerCalls
Imports minidom.Forms
Imports minidom.Anagrafica
Imports minidom.Databases
Imports minidom.Forms.Utils
Imports minidom.Finanziaria

Namespace Forms

    Partial Class Utils

        Public NotInheritable Class CCRMUtilsClas

            Public Function CreateElencoUffici(ByVal selValue As CUfficio) As String
                Dim uffici As CCollection(Of CUfficio)
                Dim writer As New System.Text.StringBuilder
                If Anagrafica.Uffici.Module.UserCanDoAction("list") Then
                    uffici = Anagrafica.Uffici.GetPuntiOperativi
                Else
                    uffici = Users.CurrentUser.Uffici
                End If
                For Each item As CUfficio In uffici
                    writer.Append("<option value=""")
                    writer.Append(GetID(item))
                    writer.Append(""" ")
                    If (GetID(selValue) = GetID(item)) Then writer.Append("selected")
                    writer.Append(">")
                    writer.Append(Strings.HtmlEncode(item.Nome))
                    writer.Append("</option>")
                Next
                Return writer.ToString
            End Function

            Public Function CreateElencoOperatori(ByVal selValue As CUser, Optional ByVal onlyValid As Boolean = True) As String
                Dim [module] As CModule
                Dim canList, canListOwn As Boolean
                Dim itemID As Integer
                Dim value As Integer
                Dim grp As CGroup
                Dim ret As New System.Text.StringBuilder
                Dim t As Boolean
                [module] = minidom.Finanziaria.Pratiche.Module
                canList = [module].UserCanDoAction("list")
                canListOwn = [module].UserCanDoAction("list_own")
                value = GetID(selValue)
                t = False
                grp = Sistema.Groups.GetItemByName("CRM")
                If canList Then
                    'Otteniamo il gruppo Finanziaria
                    If Not grp Is Nothing Then
                        For Each user As CUser In grp.Members
                            If ((onlyValid = False) OrElse (user.UserStato = UserStatus.USER_ENABLED)) Then
                                itemID = GetID(user)
                                t = t Or (value = itemID)
                                ret = ret.Append("<option value=""")
                                ret.Append(itemID)
                                ret.Append(""" ")
                                If (value = itemID) Then ret.Append("selected")
                                ret.Append(">")
                                ret.Append(Strings.HtmlEncode(user.Nominativo))
                                ret.Append("</option>")
                            End If
                        Next
                    End If
                Else
                    If Not grp Is Nothing Then
                        For Each user As CUser In grp.Members
                            If Users.CurrentUser.Uffici.SameOffice(user) Then
                                If ((onlyValid = False) OrElse (user.UserStato = UserStatus.USER_ENABLED)) Then
                                    itemID = GetID(user)
                                    t = t Or (value = itemID)
                                    ret.Append("<option value=""")
                                    ret.Append(itemID)
                                    If (value = itemID) Then ret.Append("selected")
                                    ret.Append(">")
                                    ret.Append(Strings.HtmlEncode(user.Nominativo))
                                    ret.Append("</option>")
                                End If
                            End If
                        Next
                    End If
                End If
                If (selValue IsNot Nothing) And (t = False) Then
                    itemID = GetID(selValue)
                    If value = itemID Then
                        ret.Append("<option value=""")
                        ret.Append(itemID)
                        ret.Append(""" selected>")
                        ret.Append(selValue.Nominativo)
                        ret.Append("</option>")
                    Else
                        ret.Append("<option value=""")
                        ret.Append(value)
                        ret.Append(""">INVALID (")
                        ret.Append(value)
                        ret.Append(")</option>")
                    End If
                End If

                Return ret.ToString
            End Function

            Function CreateElencoCategorie(selValue As String) As String
                Dim items As String() = {"Urgente", "Importante", "Normale", "Poco importante"}
                Return SystemUtils.CreateElenco(items, selValue)
            End Function

            Public Function GetIcon(ByVal cliente As CPersona) As String
                Dim iconURL As String = ""
                If (cliente IsNot Nothing) Then iconURL = cliente.IconURL
                If (iconURL = "") Then Return Sistema.ApplicationContext.BaseURL & "/minidom/widgets/images/default.GIF"
                Return Sistema.ApplicationContext.BaseURL & "/minidom/widgets/websvc/getthumb.aspx?p=" & iconURL & "&w=32&h=32"
            End Function

            Public Function FormatOfferta(ByVal offerta As COffertaCQS) As String
                If (offerta Is Nothing) Then Return "NULL"
                Dim testo As String = ""

                testo &= "<table>"
                testo &= "<tr>"
                testo &= "<td>"
                Dim prodotto As Finanziaria.CCQSPDProdotto = offerta.Prodotto
                If (prodotto Is Nothing) Then
                    testo &= "<b>?</b>"
                Else
                    If (prodotto.IdTipoContratto = "C") Then
                        testo &= "<b>CQS&nbsp;</b>"
                    Else
                        testo &= "<b>PD&nbsp;</b>"
                    End If
                End If
                testo &= "<span class=""blue"">" & Formats.FormatValuta(offerta.Rata) & " €</span> x <span class=""blue"">" & offerta.Durata & "</span> = " & Formats.FormatValuta(offerta.MontanteLordo) & " €"
                testo &= "</td>"
                testo &= "</tr>"
                testo &= "<tr>"
                testo &= "<td>"
                testo &= "<b>Netto:</b> <span class=""blue"">" & Formats.FormatValuta(offerta.NettoRicavo) & " €</span>, <b>TAN:</b> <span class=""blue"">" & Formats.FormatPercentage(offerta.TAN, 3) & " %</span> <b>TAEG:</b> <span class=""blue"">" & Formats.FormatPercentage(offerta.TAEG) & " %</span>"
                testo &= "</td>"
                testo &= "</tr>"
                testo &= "</table>"

                Return testo
            End Function

            Public Function GetNominativoPersona(ByVal persona As CPersona) As String
                Dim ret As String = ""
                If (persona IsNot Nothing) Then ret = persona.Nominativo
                If (ret = "") Then
                    If (persona Is Nothing) Then
                        ret = "Sconosciuto"
                    Else
                        ret = TypeName(persona) & "[" & GetID(persona) & "]"
                    End If
                End If
                Return ret
            End Function

            Public Function PreparaTestoNotifica(
                                                ByVal titolo As String,
                                                ByVal persona As CPersona,
                                                Optional ByVal operatore As CUser = Nothing,
                                                Optional ByVal data As DateTime? = Nothing
                                                ) As String
                Dim testo As String

                testo = ""
                testo &= "<table>"
                testo &= "<tr>"
                testo &= "<td><img src=""" & GetIcon(persona) & """ style=""width:32px;height:32px;"" alt=""img"" /></td>"
                testo &= "<td>"
                testo &= "<b style=""color:red;font-size:12pt;"">" & titolo & "</b></br/>"
                testo &= "<a style=""color:blue;font-size:12pt;font-weight:bold;"" href=""#"" onclick=""SystemUtils.EditItem(Anagrafica.Persone.GetItemById(" & GetID(persona) & ")); return false;"">" & Strings.HtmlEncode(GetNominativoPersona(persona)) & "<a>"
                testo &= "</td>"
                testo &= "</tr>"
                testo &= "</table>"

                If (data.HasValue) Then testo &= "<b class=""font11pt"">Data:</b> <span class""font11pt blue"">" & Formats.FormatUserDateTime(DateUtils.Now) & "</span>" & vbNewLine
                If (operatore IsNot Nothing) Then testo &= "<b class=""font11pt"">Operatore:</b> <span class""blue font11pt"">" & Strings.HtmlEncode(operatore.Nominativo) & "</span>" & vbNewLine


                Return testo
            End Function

            Public Function PreparaTestoNotificaURL(
                                               ByVal titolo As String,
                                               ByVal persona As CPersona,
                                               Optional ByVal operatore As CUser = Nothing,
                                               Optional ByVal data As DateTime? = Nothing
                                               ) As String
                Dim testo As String
                Dim baseURL As String = Sistema.ApplicationContext.BaseURL
                Dim baseName As String = "frm" & Sistema.ApplicationContext.IDAziendaPrincipale

                testo = ""
                testo &= "<table>"
                testo &= "<tr>"
                testo &= "<td><img src=""" & GetIcon(persona) & """ style=""width:32px;height:32px;"" alt=""img"" /></td>"
                testo &= "<td>"
                testo &= "<b style=""color:red;font-size:12pt;"">" & titolo & "</b></br/>"
                testo &= "<a style=""color:blue;font-size:12pt;font-weight:bold;"" target=""" & baseName & """ href=""" & baseURL & "modAnagrafica.aspx?_a=edit&id=" & GetID(persona) & """>" & Strings.HtmlEncode(GetNominativoPersona(persona)) & "<a>"
                testo &= "</td>"
                testo &= "</tr>"
                testo &= "</table>"

                If (data.HasValue) Then testo &= "<b class=""font11pt"">Data:</b> <span class""font11pt blue"">" & Formats.FormatUserDateTime(DateUtils.Now) & "</span>" & vbNewLine
                If (operatore IsNot Nothing) Then testo &= "<b class=""font11pt"">Operatore:</b> <span class""blue font11pt"">" & Strings.HtmlEncode(operatore.Nominativo) & "</span>" & vbNewLine


                Return testo
            End Function

        End Class

        Private Shared m_CRMUtils As CCRMUtilsClas = Nothing

        Public Shared ReadOnly Property CRMUtils As CCRMUtilsClas
            Get
                If (m_CRMUtils Is Nothing) Then m_CRMUtils = New CCRMUtilsClas
                Return m_CRMUtils
            End Get
        End Property

    End Class


End Namespace